File: user-service/src/main/java/io/ecommerce/userservice/error/ErrorResponse.java
package io.ecommerce.userservice.error;

import com.fasterxml.jackson.annotation.JsonFormat;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.springframework.validation.BindingResult;

import javax.servlet.http.HttpServletRequest;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * @author : choi-ys
 * @date : 2022/01/07 5:46 오후
 */
@Getter
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class ErrorResponse {

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd hh:mm:ss", timezone = "Asia/Seoul")
    private LocalDateTime timestamp;

    private String code;
    private String message;
    private String method;
    private String path;

    private List<ErrorDetail> errorDetails;

    private ErrorResponse(
            final ErrorCode errorCode,
            final HttpServletRequest httpServletRequest
    ) {
        this.code = errorCode.getCode();
        this.message = errorCode.getMessage();
        this.method = httpServletRequest.getMethod();
        this.path = httpServletRequest.getRequestURI();
        this.errorDetails = new ArrayList<>();
        this.timestamp = LocalDateTime.now();
    }

    public static ErrorResponse of(
            final ErrorCode errorCode,
            final HttpServletRequest httpServletRequest
    ) {
        return new ErrorResponse(errorCode, httpServletRequest);
    }

    private ErrorResponse(
            final ErrorCode errorCode,
            final HttpServletRequest httpServletRequest,
            final BindingResult bindingResult
    ) {
        this.code = errorCode.getCode();
        this.message = errorCode.getMessage();
        this.method = httpServletRequest.getMethod();
        this.path = httpServletRequest.getRequestURI();
        this.timestamp = LocalDateTime.now();
        this.errorDetails = ErrorDetail.of(bindingResult);
    }

    public static ErrorResponse of(
            final ErrorCode errorCode,
            final HttpServletRequest httpServletRequest,
            final BindingResult bindingResult
    ) {
        return new ErrorResponse(errorCode, httpServletRequest, bindingResult);
    }
}

File: product-service/src/main/kotlin/io/ecommerce/productgservice/error/ErrorResponse.kt
package io.ecommerce.productgservice.error

import com.fasterxml.jackson.annotation.JsonFormat
import org.springframework.validation.BindingResult
import java.time.LocalDateTime
import java.util.ArrayList
import javax.servlet.http.HttpServletRequest

/**
 * @author : choi-ys
 * @date : 2022-01-21 오전 12:41
 */
data class ErrorResponse(

    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd hh:mm:ss", timezone = "Asia/Seoul")
    val timestamp: LocalDateTime = LocalDateTime.now(),

    var code: String? = null,
    var message: String? = null,
    var method: String? = null,
    var path: String? = null,

    var errorDetails: List<ErrorDetail>? = null,
) {

    constructor(
        errorCode: ErrorCode,
        httpServletRequest: HttpServletRequest,
    ) : this() {
        code = errorCode.code
        message = errorCode.message
        method = httpServletRequest.method
        path = httpServletRequest.requestURI
        errorDetails = ArrayList()
    }


    constructor(
        errorCode: ErrorCode,
        httpServletRequest: HttpServletRequest,
        bindingResult: BindingResult,
    ) : this() {
        this.code = errorCode.code
        message = errorCode.message
        method = httpServletRequest.method
        path = httpServletRequest.requestURI
        errorDetails = ErrorDetail.of(bindingResult)
    }

    companion object {

        fun of(
            errorCode: ErrorCode,
            httpServletRequest: HttpServletRequest,
        ): ErrorResponse {
            return ErrorResponse(errorCode, httpServletRequest)
        }

        fun of(
            errorCode: ErrorCode,
            httpServletRequest: HttpServletRequest,
            bindingResult: BindingResult,
        ): ErrorResponse {
            return ErrorResponse(errorCode, httpServletRequest, bindingResult)
        }
    }

}